dtool_lookup_gui/utils/multiprocessing.py

```diff
@@ -0,0 +1,167 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+"""Multiprocessing utils."""
+
+# NOTE: depending on platform, we may have to experiment with the forking methods,
+# https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
+
+import logging
+import multiprocessing  # run task as child process to avoid side effects
+import queue
+import traceback  # forward exception from child process to parent process
+
+
+logger = logging.getLogger(__name__)
+
+
+# inspired by
+# https://stackoverflow.com/questions/19924104/python-multiprocessing-handling-child-errors-in-parent
+class Process(multiprocessing.Process):
+    """
+    Class which returns child Exceptions to Parent.
+    https://stackoverflow.com/a/33599967/4992248
+    """
+
+    def __init__(self, *args, **kwargs):
+        multiprocessing.Process.__init__(self, *args, **kwargs)
+        self._parent_conn, self._child_conn = multiprocessing.Pipe()
+        self._exception = None
+
+    def run(self):
+        try:
+            super().run()
+            self._child_conn.send(None)
+        except Exception as e:
+            tb = traceback.format_exc()
+            self._child_conn.send((e, tb))
+            raise e  # You can still rise this exception if you need to
+
+    @property
+    def exception(self):
+        if self._parent_conn.poll():
+            self._exception = self._parent_conn.recv()
+        return self._exception
+
+
+class StatusReportingChildProcessBuilder:
+    """Outsource serial functions with status report handlers.
+
+    The status report handler is expected to conform to the
+    click.ProgressBar interface. In particular, it must exhibit an
+    update(val) method.
+
+    For any function that runs serial and reports status via such a callback,
+    this wrapper can run them in a non-blocking forked process and forward the
+    status reports via queue to the callback.
+
+    The function must have the signature
+
+        func(*args, status_report_callback=None, stop_event_callback=None)
+
+    and need not make use of any of the two callback functions. It can check on
+    the stop_event_callback() to finish gracefully when the parent process wants
+    the child process to finish.
+    """
+    def __init__(self, target, status_report_callback):
+        self._target = target
+        self._status_report_handler = status_report_callback
+
+    def __call__(self, *args):
+        """Spawn child process to assure my environment stays untouched."""
+        return_value_queue = multiprocessing.Queue()
+        status_progress_queue = multiprocessing.Queue()
+        stop_event = multiprocessing.Event()
+        process = Process(target=self.target_wrapper, args=[return_value_queue, status_progress_queue, stop_event, *args])
+        process.start()
+
+        # wait for child to queue objects and
+        # check whether child raises exception
+        while return_value_queue.empty():
```

Comment:
I don't understand this. Doesn't this while loop block?

Reply:
It shouldn't, the only blocking lines should be those at the bottom, https://github.com/IMTEK-Simulation/dtool-lookup-gui/blob/bd0e9b33040316749f7db1d3f3e36f0b610f482e/dtool_lookup_gui/utils/multiprocessing.py#L125-L128.

Reply:
Don't you need to poll the queue regularly to get the progress information? How does that work?

Reply:
There are two queues, one for the progress and one only for the return value of the function (if it has any). The two polling calls within the loop https://github.com/IMTEK-Simulation/dtool-lookup-gui/blob/f55e6bca037b9a179070308f2de20999a09a3e29/dtool_lookup_gui/utils/multiprocessing.py#L101 and https://github.com/IMTEK-Simulation/dtool-lookup-gui/blob/f55e6bca037b9a179070308f2de20999a09a3e29/dtool_lookup_gui/utils/multiprocessing.py#L109 don't wait. The loop itself of course was blocking, sorry, see below ...

Reply:
Okay, now it makes sense